[clangd] Add inlay hints for default function arguments

torshepherd opened the pull request:
Title. This PR adds support for defaulted function arguments:

![image](https://github.com/llvm/llvm-project/assets/49597791/2a6b83a9-b918-4363-be67-d79ce244b067)

as inlay hints.

If the list of default args is too long (based on TypeNameLimit), shows "..." instead:

![image](https://github.com/llvm/llvm-project/assets/49597791/c48acd28-091a-4d7d-9344-01e12c622024)

Note: I wrote this before https://github.com/llvm/llvm-project/pull/85497 landed. With this we should probably also follow up with go-to-definition on the lambda captures at least; I could see that being helpful

Edit (9/23): Removed ' and implicit lambda captures'

zyn0217 reviewed (commented):
My two cents:

While I appreciate the value of inspecting more semantic information about lambdas, I still have mixed feelings about user experience.

Admittedly, presenting implicitly captured variables provides users with insight into ODR-used variables, but I'm unsure about our strategy: We would probably end up showing too many hints inside the lambda capture, which would visually elongate the lambda header and distract users from reading.

Indeed, we have a configuration entry to restrict the length, but I feel that's insufficient (Although we're baking it with TypeNameLimit currently, which is also inappropriate). For example, a user might have a lambda capturing several variables named in single letters e.g. a-z. As a result, we probably eventually display `[&: a, b, c, d, e, f, g, ...](...)`, which is IMO too spammy.

That said, I wonder if we can implement the capture stuff in hovers instead. From what I can think of, we present something e.g.
```cpp
// implicitly captured by reference/copy:
int a;
std::string s;
...
```
when user hovers over the capture defaults.

(I don't have a strong opinion on the hints for default arguments. It would also be great to hear @HighCommander4 and @kadircet's input.)

torshepherd commented on clang-tools-extra/clangd/unittests/InlayHintTests.cpp:
```diff
@@ -15,9 +15,12 @@
+#include "llvm/Support/raw_ostream.h"
 #include "gmock/gmock.h"
 #include "gtest/gtest.h"
+#include <optional>
 #include <string>
+#include <utility>
```
I believe std::move - let me check if they're still used in the latest revision though

torshepherd replied:
Yep, move

torshepherd commented on clang-tools-extra/clangd/unittests/InlayHintTests.cpp:
```diff
@@ -1458,13 +1463,66 @@ TEST(TypeHints, DefaultTemplateArgs) {
     auto $var[[var]] = foo();
-    A<float> bar[1];
+    A<float> baz;
+    A<float> bar[1]{};
```
My fault, an artifact from a different set of inlay hints I decided not to include (show implicit default constructions with '{}')

torshepherd commented on clang-tools-extra/clangd/unittests/InlayHintTests.cpp:
```diff
@@ -1568,7 +1626,7 @@ TEST(TypeHints, SubstTemplateParameterAliases) {
 
   llvm::StringRef VectorIntPtr = R"cpp(
-    vector<int *> array;
+    vector<int *> $init[[array]];
```
Another artifact, sorry

torshepherd commented on clang-tools-extra/clangd/ConfigCompile.cpp:
```diff
@@ -504,10 +503,10 @@ struct FragmentCompiler {
-                 "Latency of clang-tidy check '{0}' is not known. "
-                 "It will only run if ClangTidy.FastCheckFilter is Loose or None",
-                 Str)
+             llvm::formatv("Latency of clang-tidy check '{0}' is not known. "
+                           "It will only run if ClangTidy.FastCheckFilter is "
+                           "Loose or None",
+                           Str)
```
Sorry, I think this was the formatter messing with it

torshepherd commented on clang-tools-extra/clangd/InlayHints.cpp:
```diff
@@ -372,6 +382,34 @@ maybeDropCxxExplicitObjectParameters(ArrayRef<const ParmVarDecl *> Params) {
+    return Capture.getCapturedVar()->getName();
+  if (Capture.capturesThis())
+    return llvm::StringRef{"this"};
+  return llvm::StringRef{"unknown"};
```
Changed this to a switch case

torshepherd commented on clang-tools-extra/clangd/InlayHints.cpp:
```diff
@@ -372,6 +382,34 @@ maybeDropCxxExplicitObjectParameters(ArrayRef<const ParmVarDecl *> Params) {
+    if (!IsFirst)
+      Out.append(", ");
+    else
+      IsFirst = false;
```
Changed this to use raw_string_ostream, lmk how it looks

torshepherd commented:
Ok I took a stab at doing lambda captures as hover instead.

- Targeting the exact '=' or '&' is nontrivial since the capture list isn't in the AST, the lambda is just `LambdaExpr` with children for the parameters and the CompoundStmt body.
- `auto MyLambda = [...` already produces a nice function-like hover over the variable. I tested adding a "Captures:" bullet-list, it works well enough but isn't very noticeable.
- We could take the above HoverInfo over the variable containing the lambda and apply it to hovering the lambda itself as well?


It strikes me, if we could either add HoverInfo or styling to the variables within the body of the lambda themselves, that would be the most readable representation probably.

For instance:

- Can we add an inlay hint like `[&] ` before actual variables in the body of the lambda?
- Can we apply a semantic token to style lambda captures differently? Maybe as `property` since they are similar conceptually to struct members?
- Can we use the `mutable`-ness of the lambda to color tokens as const if they are captured by copy and the lambda is not mutable?

Overall I think capture hints should be removed from this PR so we can get default args in in the meantime.

HighCommander4 reviewed (changes requested):
Apologies for the slow response time, it's been quite a busy summer.

High-level feedback:
 * Let's definitely split the default arguments part from the lambda captures part. Even if we decide we want both, it helps to have separate features in separate patches, so that e.g. if one causes a problem and needs to be reverted, it's easy to do that without affecting the other.
 * I tend to agree with @zyn0217 that lambda captures might be a better fit for hovers than inlay hints. I think we could solve the problem targeting the capture-default token: if `SelectionTree` selects the `LambdaExpr`, massage the input position into a `SourceLocation`, compare it to `getCaptureDefaultLoc()`, and if so then take a different branch (in `Hover.cpp` code). (This doesn't prevent us from potentially also pursuing the idea of adding a "captures" section to the hover for the lambda variable.)
 * For the default argument hints, what do you think about using `name: value` syntax instead of `name = value` syntax, for consistency with parameter hints?

torshepherd committed:
just defaults

torshepherd commented on clang-tools-extra/clangd/InlayHints.cpp:
```diff
@@ -709,7 +762,8 @@ class InlayHintVisitor : public RecursiveASTVisitor<InlayHintVisitor> {
   using NameVec = SmallVector<StringRef, 8>;
 
-  void processCall(Callee Callee, llvm::ArrayRef<const Expr *> Args) {
+  void processCall(Callee Callee, SourceRange LParenOrBraceRange,
```
Sorry yep that should be RParenOrBraceRange

torshepherd commented on clang-tools-extra/clangd/InlayHints.cpp:
```diff
@@ -372,6 +383,38 @@ maybeDropCxxExplicitObjectParameters(ArrayRef<const ParmVarDecl *> Params) {
+    return llvm::StringRef{"this"};
+  case LCK_ByCopy:
+  case LCK_ByRef:
+  case LCK_VLAType:
```
Resolving since I removed LambdaCaptures from this PR

torshepherd commented on clang-tools-extra/clangd/InlayHints.cpp:
```diff
@@ -755,12 +812,33 @@ class InlayHintVisitor : public RecursiveASTVisitor<InlayHintVisitor> {
       bool NameHint = shouldHintName(Args[I], Name);
       bool ReferenceHint = shouldHintReference(Params[I], ForwardedParams[I]);
 
+      bool IsDefault = isa<CXXDefaultArgExpr>(Args[I]);
```
Must be because `shouldHintName` and `shouldHintReference` are both false. Let me double check

torshepherd replied:
Ah. It's actually because `getHintRange` returns a nullopt. I added a guard to the if statement to prevent regressions

torshepherd commented:
Hmm another case to consider:

![image](https://github.com/user-attachments/assets/193c9570-0e6b-47cd-8d8c-0279e727eac2)

In this case the argument is unnamed. I think for this I should just insert `/* unused */` or `/* unnamed */` ?

torshepherd committed:
Fix hints for unnamed

torshepherd committed:
fixes

torshepherd commented:
> Hmm another case to consider:
> 
> ![image](https://private-user-images.githubusercontent.com/49597791/372187742-193c9570-0e6b-47cd-8d8c-0279e727eac2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc3MjA2MDUsIm5iZiI6MTcyNzcyMDMwNSwicGF0aCI6Ii80OTU5Nzc5MS8zNzIxODc3NDItMTkzYzk1NzAtMGU2Yi00N2NkLThkOGMtMDI3OWU3MjdlYWMyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTMwVDE4MTgyNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTVjMTRkNDM4YWFlYjlhN2UzMjc5YzhlMzg1ZmQ1ZTJmZDgzMzJiMGIyZjAwMmJiODViOWU3NjU0NWExZDYxYzImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.1oEmnUxc-hEikrOjLY9iZvGMqaQp8acwcrfB1bDayT0)
> 
> In this case the argument is unnamed. I think for this I should just insert `/* unused */` or `/* unnamed */` ?

Fixed this and added a test case

![image](https://github.com/user-attachments/assets/49e29147-a2ed-4482-b2d5-d0b2996fdfd2)

torshepherd committed:
if

torshepherd commented:
Ping

HighCommander4 commented on clang-tools-extra/clangd/InlayHints.cpp:
```diff
@@ -709,7 +738,8 @@ class InlayHintVisitor : public RecursiveASTVisitor<InlayHintVisitor> {
   using NameVec = SmallVector<StringRef, 8>;
 
-  void processCall(Callee Callee, llvm::ArrayRef<const Expr *> Args) {
+  void processCall(Callee Callee, SourceRange RParenOrBraceRange,
```
I think it would be clearer for the `SourceRange RParenOrBraceRange` parameter to be `SourceLocation RParenOrBraceLoc` instead.

Both call sites in fact pass a `SourceLocation`, which currently gets implicitly converted to a singleton range.

I'd rather we do that conversion inside the function's implementation, and make it explicit with `SourceRange{RParenOrBraceLoc}`.

HighCommander4 commented on clang-tools-extra/clangd/InlayHints.cpp:
```diff
@@ -372,6 +381,25 @@ maybeDropCxxExplicitObjectParameters(ArrayRef<const ParmVarDecl *> Params) {
 }
 
+template <typename R, typename P>
+std::string joinAndTruncate(R &&Range, size_t MaxLength,
```
Now that the function has only one call site, can we drop the `GetAsStringFunction` parameter and just assume the elements are themselves strings?

HighCommander4 commented on clang-tools-extra/clangd/unittests/InlayHintTests.cpp:
```diff
@@ -1465,6 +1469,34 @@ TEST(TypeHints, DefaultTemplateArgs) {
+    void baz(int = 5) { if (false) baz($unnamed[[)]]; };
+  )cpp";
+
+  assertHints(InlayHintKind::DefaultArgument, Code,
```
It would be nice to have one check where the token the hint is attached to is an R-brace rather than an R-paren

HighCommander4 reviewed (commented):
Thanks for the update!

The patch is looking pretty good, just some minor comments remaining and then it should be good to go.

torshepherd committed:
Comments and partial refactor

torshepherd commented:
Ok I found a couple of nontrivial issues that I took care of in the latest commit.

1. Using `DefaultArguments` required having parameter names set to true as well. I separated them so that both have a modality that works with or without the other. See screenshot for defaults w/out parameter names:

![image](https://github.com/user-attachments/assets/fb5a9608-33aa-41bb-bf2e-1fc4361ca0da)

2. Having newlines in the default definition caused `LF` to be displayed in the hints. I fixed this by just displaying `...` any time a default definition contains `\n`.
3. I was getting `Property DefaultArguments` is not allowed in `config.yaml`. Is this a schema issue? I wasn't able to find where to update this

Also, a couple of follow-up things I should do:
- Add support for default `Designators`. For instance,

```cpp
struct Foo { int a; int b = 2 };
Foo f{1}; // Currently shows 'Foo f{.a=1}', should show 'Foo f{.a=1, .b=2}'
```
- Add support for template args? such as `std::vector<int, std::allocator<int>>` when user types `std::vector<int>`. This could also fall under "add InlayHints for FTAD and CTAD", which may be a nongoal, idk
- Add a link to the source location of the definition (leveraging https://github.com/llvm/llvm-project/pull/85497)
- Do the formatting better. Basically, in some cases, collapsing newlines preserves the meaning of the code and therefore we should do that. However in other cases (comments for example), collapsing newlines makes the code gibberish (`3 // + 2` from the test case for example), and we should just show `...` instead.

HighCommander4 commented:
>    3. I was getting `Property DefaultArguments` is not allowed in `config.yaml`. Is this a schema issue? I wasn't able to find where to update this

That sort of diagnostic is likely produced by a YAML plugin, which uses a schema from https://github.com/SchemaStore/schemastore.

I do file a SchemaStore issue once per clangd release about updating the clangd schema, here's the one for clangd 19: https://github.com/SchemaStore/schemastore/issues/4022.

If you'd like to file one about `DefaultArguments` sooner, or even submit a PR, please feel free.

HighCommander4 commented on clang-tools-extra/clangd/unittests/InlayHintTests.cpp:
```diff
@@ -1497,6 +1497,47 @@ TEST(DefaultArguments, Smoke) {
+    };
+    struct Foo {
+      Foo(int, Baz baz = //
+              Baz{$unnamed[[}]]
```
nit: unnamed --> abbreviated? (the parameter itself is named here)

HighCommander4 reviewed (approved):
Thanks for the update and for spotting and fixing the added issues. I agree that keeping `DefaultArguments` orthogonal to `Parameters` is a good choice.

torshepherd committed:
comment

torshepherd committed:
Relnotes

torshepherd commented:
Accepted the comment and added relnotes. Lmk if you need anything else prior to merge

HighCommander4 commented:
Thanks. The buildkite failure looks unrelated (it's a `clang-move` test, and this patch only touches `clangd` code), so I will go ahead and merge this.
